pslab/bus/i2c.py

```diff
@@ -136,10 +158,273 @@ def _status(self):
+        :meth:`I2CSlave.write`.
+
+        Returns
+        -------
+        data : int
+            A byte interpreted as a uint8.
+        """
+        self._device.send_byte(CP.I2C_HEADER)
+        self._device.send_byte(CP.I2C_READ_MORE)
+        data = self._device.get_byte()
+        self._device.get_ack()
+
+        return data
+
+    def _read_end(self) -> int:
+        """Read data from I2C then send NACK.
+
+        The I2C bus needs to be initialized and set to the correct slave address first.
+
+        It is a primitive I2C method, prefered to use :meth:`I2CSlave.read` and
+        :meth:`I2CSlave.write`.
+
+        Returns
+        -------
+        data : int
+            A byte interpreted as a uint8.
+        """
+        self._device.send_byte(CP.I2C_HEADER)
+        self._device.send_byte(CP.I2C_READ_END)
+        data = self._device.get_byte()
+        self._device.get_ack()
+
+        return data
+
+    def _read(self, bytes_to_read: int) -> bytearray:
+        """Read data from I2C.
+
+        The I2C bus needs to be initialized and set to the correct slave address first.
 
-class I2CSlave:
+        It is a primitive I2C method, prefered to use :meth:`I2CSlave.read` and
+        :meth:`I2CSlave.write`.
+
+        Parameters
+        ----------
+        bytes_to_read : int
+            Number of bytes to read from I2C.
+
+        Returns
+        -------
+        data : bytearray
+            Read data as a bytearray.
+        """
+        data = bytearray()
+
+        for _ in range(bytes_to_read - 1):
+            data.append(self._read_more())
+        data.append(self._read_end())
+
+        return data
+
+    def _read_bulk(
+        self, address: int, bytes_to_read: int, register_address: int = 0x0
+    ) -> bytearray:
+        """Read data from I2C device.
+
+        This method relies on the slave device auto incrementing its internal
+        pointer after each read. Most devices do, but it is not part of the
+        I2C standard. Refer to the device's documentation.
+
+        It is a primitive I2C method, prefered to use :meth:`I2CSlave.read` and
+        :meth:`I2CSlave.write`.
+
+        Parameters
+        ----------
+        address : int
+            7-bit I2C device address.
+        bytes_to_read : int
+            Number of bytes to read from slave device.
+        register_address : int, optional
+            Slave device internal memory address to read from. The default
+            value is 0x0.
+
+        Returns
+        -------
+        data : bytearray
+            Read data as a bytearray.
+        """
+        self._device.send_byte(CP.I2C_HEADER)
+        self._device.send_byte(CP.I2C_READ_BULK)
+        self._device.send_byte(address)
+        self._device.send_byte(register_address)
+        self._device.send_byte(bytes_to_read)
+        data = self._device.read(bytes_to_read)
+        self._device.get_ack()
+
+        return bytearray(data)
+
+
+class I2CMaster(I2CPrimitive):
```

Comment:
Does I2CMaster actually need to inherit from I2CPrimitive? Doesn't look like it uses anything from it, as far as I can tell.